Support publishing of multi-architecture OCI container images

surayya-MS opened the pull request:
### Context
Fixes https://github.com/dotnet/sdk-container-builds/issues/87. Currently, customers use our tooling to publish single architectural container images providing architecture via `RuntimeIdentifier` or `ContainerRuntimeIdentifier` properties. It's not possible to publish multi-architectural container image. This change allows customers to do that.

### Customer impact
With this change customers will be able to publish multi-arch container images to remote registries, or publish single-arch images for each architecture locally (either directly to docker/podman or as tarballs).

### Details
It is not possible to create multi-arch image in `docker` because the individual images for each architecture must be available in the remote registry. For `podman` it is possible, however, that is not part of these PR. It will be the next iteration.

### Changes
1. `PublishContainer` target renamed to `_PublishSingleContainer`
2.  New `_PublishMultiArchContainers` target that calls `_PublishSingleContainer` for each arch and then creates image index
3.  `PublishContainer` target calls either `_PublishSingleContainer` or `_PublishMultiArchContainers`  depending on a condition
4.  Implemented new task `CreateImageIndex` that creates the image index (manifest list it's same thing) and pushes to the remote registry.

### Testing
1. Added unit and integration tests
2. Tested manually


### Risk
_Low_ - the existing tests ensure that other scenarios were not affected.

baronfel committed:
add code to containers check to allow users to silence it

baronfel committed:
teeny tiny amount of work to show an example

surayya-MS committed:
implement CreateImageIndex task

surayya-MS committed:
fix

surayya-MS committed:
rename

surayya-MS committed:
remove comment

surayya-MS committed:
fix

surayya-MS committed:
remove usings

surayya-MS committed:
fix

surayya-MS committed:
fix

surayya-MS committed:
fix

surayya-MS committed:
Revert "fix"

This reverts commit 7521fa28ef0a2ef967d7d512e9ed082dbe2ccf53.

surayya-MS committed:
Revert "fix"

This reverts commit ec4c94f5c9b17cf7feaa3f1543862c45fc9a26af.

surayya-MS committed:
fix

surayya-MS committed:
Merge branch 'release/8.0.4xx' into multi-arch-containers

surayya-MS committed:
fix init Registry in CreateImageIndex

surayya-MS committed:
fix condition for calling _PublishMultiArchContainers or _PublishSingleContainer

surayya-MS committed:
small fix

surayya-MS committed:
fix condition for including rids in _PublishMultiArchContainers

surayya-MS committed:
Revert "fix condition for including rids in _PublishMultiArchContainers"

This reverts commit 905039de9b9e360369b8fe74fee1872f070ad443.

surayya-MS committed:
optimize targets order

surayya-MS committed:
Merge branch 'release/8.0.4xx' into multi-arch-containers

baronfel commented:
A few notes that we will need to handle based on my trials at https://github.com/baronfel/sdk-container-demo/pull/15

* The `ContainerRepository` property is only inferred for single-RID scenarios, not multi-RID so the CreateImageIndex task crashes. 
  * We should be able to infer this property similarly to how we infer it for the single-RID publishes
* `ContainerImageTags` needs to be supported for the inner builds - when `ContainerImageTags` is specified we get crashes on the [single-image Publishes](https://github.com/dotnet/sdk/pull/43085/files#diff-a5f4ada40b043c8a36b7a1d56148dcfd0abcfe13fa82dbfb456f7c43bebedff7R301) because the single-image publishes unconditionally forward along a single tag.
  *  Instead of always forcing a single ContainerImageTag, we should append the RID to each ContainerImageTags value if specified
  * The CreateImageIndex Task should also accept ContainerImageTags similar to the single-image push
* Using the package is rough right now - the in-SDK checks complain at you. In addition, if you have both the SDK and the package the SDK's `Import` always overrides the package. I had to work around this [here](https://github.com/baronfel/sdk-container-demo/pull/15/files#diff-50e91c82311ea26f2a73202525dfdf2b0a89c178ee666b2bf3ad4c84ac4c06dcR1-R11)
  * We should set the `_ContainersTargetsDir` property to point to the containers package path in the package .props so that users don't have to do all of this

surayya-MS committed:
create each image and skip image index when no container registry is provided

surayya-MS committed:
Merge branch 'release/8.0.4xx' into multi-arch-containers

surayya-MS committed:
call ComputeContainerImageTag target once for all containers

surayya-MS committed:
Revert "call ComputeContainerImageTag target once for all containers"

This reverts commit b52a15fbd1aa6f9dcc8b00019ede07501b840809.

surayya-MS committed:
Reapply "call ComputeContainerImageTag target once for all containers"

This reverts commit d9a3969b851f92ef5f4d65fd65d849c9bc1ef4be.

surayya-MS committed:
call ComputeContainerBaseImage and ComputeContainerConfig targets once for all containers

surayya-MS committed:
fix _SkipCreateImageIndex condition

surayya-MS committed:
support ContainerImageTags for image index

surayya-MS committed:
improved logging for CreateImageIndex

surayya-MS committed:
support posting tarballs to ArchiveRegistry for multi-arch scenario

surayya-MS committed:
small fix

baronfel commented on src/Containers/packaging/build/Microsoft.NET.Build.Containers.targets:
```diff
@@ -272,6 +300,92 @@
+          _ContainerAppCommandArgs=@(ContainerAppCommandArgs, ';');
+          ContainerAppCommandInstruction=$(ContainerAppCommandInstruction);
+          _ContainerDefaultArgs=@(ContainerDefaultArgs, ';');
+          _ContainerLabel=@(ContainerLabel, ';');
+          _ContainerPort=@(ContainerPort, ';');
+          _ContainerEnvironmentVariables=@(ContainerEnvironmentVariables, ';');
```
These three Items can't be 'passed' into the inner build like this because they contain metadata that gets removed when they are converted back to properties. As a result, when parsed back into Items during `_ParseParametersForPublishingSingleContainer` the Items exist but no longer have any of their useful values.

@rainersigwald is there a better pattern for computing properties and Items that should be provided to an Inner build?

rainersigwald replied:
There's not really a good "push" model for items with metadata. You can "pull" via the return value of a target called with MSBuild; that might be an option here (to have the inner builds call back to the outer build to get the info). Otherwise you have to pass structured data and deserialize it on the other side (JSON or something).

baronfel replied:
this is an interesting question though - you could see for some of these (env var/arguments) might need to vary per-RID.

baronfel commented:
I tested the next round of changes and it's looking really good!

* ContainerRepository now works for the outer build
* ContainerImageTags is working as expected for the outer and inner builds
* The 'skip publishing an Image Index' is working great when publishing multiple images to local sources (Podman manifest support will come later)

I noted two problems to dig into:
* The labels, env vars, and ports are not correctly being passed into the inner builds because data is lost in the current transmission format. I don't know a good answer for this off the top of my head, so I pinged Rainer for guidance.
* the generated manifests use the docker distribution list media type - but this needs to be configurable. The specs for each of the two 'list' types declare what they allow:
  * Docker Manifest Lists allow `application/vnd.docker.distribution.manifest.v2+json` and `application/vnd.docker.distribution.manifest.v1+json`
  * OCI Image Indexes use `[application/vnd.oci.image.manifest.v1+json](https://github.com/opencontainers/image-spec/blob/main/manifest.md)`
  * so we will need to pick one based on what the `mediaTypes` of the individual base images are

surayya-MS committed:
import containers targets only when the package was not referenced

surayya-MS committed:
set _ContainersTargetsDir to point to the containers package path

surayya-MS committed:
rename a property

surayya-MS committed:
fix e2e tests

surayya-MS committed:
1. group outputs of CreateNewImage task into GeneratedContainers
2. set mediaType of each image from manifest in CreateImageIndex
3. set arch and os from config in CreateImageIndex

marcpopMSFT commented:
@dotnet/sdk-container-builds-maintainers Old PR triage: Another 8.0.Nxx targeted container PR. Please review and determine if it should be brought forward or closed.

baronfel commented:
This one is the multi-arcg work for our internal partners that we are targeting for end of the month.

surayya-MS committed:
Merge branch 'release/8.0.4xx' into multi-arch-containers

surayya-MS committed:
set image index mediaType accurately

surayya-MS committed:
refactor CreateImageIndex

surayya-MS committed:
change ManifestV2.MediaType to optional

surayya-MS committed:
initial oci tarball fix

surayya-MS committed:
set image name and tag for oci tarball publishing

surayya-MS committed:
make Manifesr.MediaType optional

surayya-MS committed:
Revert "change ManifestV2.MediaType to optional"

This reverts commit 5c2b717254255ed624a195f92a2a09fc11ce73d5.

surayya-MS committed:
small fix

surayya-MS committed:
add annotations to ImageIndexV1 entries; fix oci tarball publishing

surayya-MS committed:
refactor WriteDockerImageToStreamAsync and WriteOciImageToStreamAsync

surayya-MS committed:
fix CreateImageIndex because of added annotations to oci image index entries

surayya-MS committed:
small fix

surayya-MS committed:
log error when trying to create tarball for oci image with multiple tags

surayya-MS committed:
improve CreateImageIndex

surayya-MS committed:
remove usings

surayya-MS commented:
1. Fixed OCI image tarball publishing (for single arch) - currently tarballs for oci image have docker structure
3. Added support for creating OCI image index for multi-arch

baronfel committed:
BugFix: UseAppHost is the property that controls if a project should be launched with the binary or through dotnet

baronfel committed:
Bugfix: Ensure that per-RID publishing occurs so that the per-RID assets are bundled into the per-RID application containers.

surayya-MS committed:
small fix

surayya-MS committed:
refactor CreateImageIndex

surayya-MS committed:
small fix

surayya-MS committed:
small fix

surayya-MS committed:
change error messages

surayya-MS committed:
add ImageIndexGeneratorTests

surayya-MS committed:
remove usings

surayya-MS committed:
revert "make Manifesr.MediaType optional"

surayya-MS committed:
merge main

surayya-MS committed:
fix

surayya-MS committed:
validate images items metadata in CreateImageIndex; rename field

surayya-MS committed:
fix

surayya-MS committed:
remove not used properties

surayya-MS committed:
add output GeneratedContainerMediaType to CreateNewImage; use it in CreateImageIndex

surayya-MS committed:
add tests that check docker and oci tarball correct structure

surayya-MS committed:
fix tests

surayya-MS committed:
fix tests

surayya-MS committed:
fix oci tarball tests

surayya-MS committed:
add validation for TargetRuntimeIdentifiers in ComputeDotnetBaseImageAndTag task

surayya-MS committed:
fix per image publish when  rid is linux-musl

surayya-MS committed:
Revert "fix per image publish when  rid is linux-musl"

This reverts commit d182cfaf870d858e576404c7b236766f58c5e838.

surayya-MS committed:
Reapply "fix per image publish when  rid is linux-musl"

This reverts commit 8a0dc7b5355ca055c1f349b17ccfb6baa7a16dce.

surayya-MS committed:
fix setting RuntimeIdentifier without musl

surayya-MS committed:
Revert "fix setting RuntimeIdentifier without musl"

This reverts commit 44181534434ba4dcebc1b571c852f4a18dacbe02.

surayya-MS committed:
Revert "Reapply "fix per image publish when  rid is linux-musl""

This reverts commit c2a29c44b8510a9c8bd0610b7de14e58ac769bcf.

surayya-MS committed:
use BuiltImage.ManifestMediaType to create tarballs

surayya-MS committed:
add tests for multiple rids in ComputeDotnetBaseImageAndTag.TargetRuntimeIdentifiers

surayya-MS committed:
add CreateImageIndex baseline test

surayya-MS committed:
small fix

surayya-MS committed:
fix type and comment

surayya-MS committed:
add e2e tests for multi-arch scenario

surayya-MS committed:
fixed passing ContainerEnvironmentVariables in _PublishMultiArchContainers target

surayya-MS committed:
added test for env vars for multi-arch

surayya-MS committed:
fix by adding a condition

surayya-MS committed:
Merge branch 'release/8.0.4xx' into multi-arch-containers

surayya-MS committed:
fixed passing ContainerPort

surayya-MS committed:
add test for ports

surayya-MS committed:
improve env var e2e test

surayya-MS committed:
improve ports test

surayya-MS committed:
improve e2e muclti-arch tests

surayya-MS committed:
fixed passing labels for multi-arch

surayya-MS committed:
added test for labels

surayya-MS committed:
simplify passing env vars, ports and labels

surayya-MS committed:
set --platform option in docker command in e2e tests

surayya-MS committed:
merge from main

surayya-MS committed:
remove usings

surayya-MS committed:
remove line

surayya-MS committed:
try fixing tests

surayya-MS committed:
Revert "try fixing tests"

This reverts commit ece7ee2cead9c08eff6dccf9a1c79d71b3be7946.

surayya-MS committed:
fix tests

surayya-MS committed:
fix BuildingImageIndex message

surayya-MS committed:
move target position in the file

surayya-MS committed:
change order of msbuild task targets in _PublishMultiArchContainers target

surayya-MS committed:
better name for target _ParseItemsForPublishingSingleContainer

surayya-MS committed:
change env vars and ports e2e tests to be multi-rid

surayya-MS committed:
Merge branch 'release/8.0.4xx' into multi-arch-containers

baronfel commented on src/Containers/packaging/build/Microsoft.NET.Build.Containers.props:
```diff
@@ -15,6 +15,7 @@
 
     <!--Register our custom task-->
     <UsingTask TaskName="$(MSBuildThisFileName).Tasks.CreateNewImage" AssemblyFile="$(ContainerCustomTasksAssembly)"/>
+    <UsingTask TaskName="$(MSBuildThisFileName).Tasks.CreateImageIndex" AssemblyFile="$(ContainerCustomTasksAssembly)"/>
```
@surayya-MS have you tried publishing a container from VS? I worry that this using task will only load a .NET version of the Task (even if the CreateImageIndex Task is never called from the single-RID VS publishing workflow).

surayya-MS replied:
I tried it out and it doesn't work. Details:
1. I added Docker support to the project via VS UI, and it added `RuntimeIdentifiers` to my project. 
2. When I ran the project via VS UI, I got error
```Severity	Code	Description	Project	File	Line	Suppression State	Details
Error (active)	MSB4006	There is a circular dependency in the target dependency graph involving target "Publish". Since "_CopyAspNetCoreFilesToIntermediateOutputPath" has "DependsOn" dependence on "Publish", the circular is "Publish<-_CopyAspNetCoreFilesToIntermediateOutputPath<-CorePublish<-DotNetPublish<-_PublishBuildAlternative<-Publish".	container-webapp	C:\Program Files\dotnet\sdk\8.0.405-dev\Sdks\Microsoft.NET.Sdk.Publish\targets\CopyTargets\Microsoft.NET.Sdk.Publish.CopyFiles.targets	81		
```
3. I figured out it might be the `Publish` target that we add to the `MSBuild` task before `_PublishSingleContainer`. I removed `Publish` target to test my theory, and got another error saying that the image {my-app-name}:latest was not found. The image was created with another tag `latest-linux-x64`. So, VS relies on the tag being `latest`.

We need extra work on VS side to support multi-arch publishing

surayya-MS replied:
As discussed, added check for `BuildingInsideVisualStudio` to call `_PublishSingleContainer` for fixing publish container from VS. Manually tested

baronfel commented on src/Tests/Microsoft.NET.Build.Containers.IntegrationTests/CreateImageIndexTests.cs:
```diff
@@ -0,0 +1,169 @@
+    }
+
+    [DockerAvailableFact]
+    public async Task CreateImageIndex_Baseline()
```
This test covers the fact that we can use the CreateImageIndex Task to create a manifest, but it does the 'orchestration' manually. Can you 

* add a test that that just calls the 'normal' entrypoint and verifies that the multi-rid orchestration does happen for publishing the index to a registry?
* add a test that that just calls the 'normal' entrypoint and verifies that the multi-rid orchestration does happen for publishing to tarballs?

These are the two major use cases that our partner team has.

surayya-MS replied:
Added end to end tests EndToEndMultiArch_LocalRegistry, EndToEndMultiArch_ArchivePublishing, EndToEndMultiArch_Registry

baronfel committed:
Merge branch 'release/8.0.4xx' into multi-arch-containers

surayya-MS committed:
add e2e test for multi-arch tarballs publishing

surayya-MS committed:
Merge branch 'multi-arch-containers' of https://github.com/surayya-MS/sdk into multi-arch-containers

surayya-MS committed:
rename test

surayya-MS committed:
fix publishing container vis VS

surayya-MS committed:
Merge branch 'release/8.0.4xx' into multi-arch-containers

baronfel commented:
For posterity - the red template engine test is being merged on red on 8.x because the temporary fix we've done on main won't be backported.

SimaTian commented on src/Containers/Microsoft.NET.Build.Containers/ImageIndexGenerator.cs:
```diff
@@ -0,0 +1,128 @@
+
+        string firstManifestMediaType = imageInfos[0].ManifestMediaType;
+
+        if (firstManifestMediaType == SchemaTypes.DockerManifestV2)
```
How likely are we to support more manifest types in the future? (I don't know if they exist and/or if they are reasonable to be supported)
Probably not too relevant right now, but switching as future proofing might be reasonable here.

surayya-MS replied:
Not likely. These are the ones that exist today, to my knowledge

surayya-MS replied:
that is also the reason for not going with "switch" - to emphasize there likely won't be others

surayya-MS committed:
Merge branch 'release/8.0.4xx' into multi-arch-containers

SimaTian commented on src/Containers/Microsoft.NET.Build.Containers/ImageIndexGenerator.cs:
```diff
@@ -0,0 +1,128 @@
+        }
+    }
+
+    private static (string, string) GenerateDockerManifestList(ImageInfo[] images, string firstManifestMediaType)
```
This function is almost identical to the `GenerateDockerManifestList` one.
Would merging the two either by templating them or by switching the type make sense?
```private static (string, string) GenerateManifestList<TManifest, TManifestList> (ImageInfo[] Images, SchemaType mediaType) { }...
```
also, we can get the firstManifestMediaType in the function (or infer it by the type of the function) so passing it as a parameter looks somewhat redundant.

surayya-MS replied:
I agree the code repetition looks bad here. The problem here is that `PlatformSpecificManifest` and `ManifestListV2` (similarly in `GenerateOciImageIndex` - `PlatformSpecificOciManifest` and `ImageIndexV1`) are not classes but structs, so they cannot be united by one parent. `PlatformSpecificManifest` and `PlatformSpecificOciManifest` have a slight difference in `annotations` but I don't need this in here, and I could use potentially `ManifestListV2` for both and shorten the code.  I made a choice to be more explicit here and state exactly what struct we are building - with a downside being the code duplication.
However, I don't feel strongly about this choice. I will leave this conversation unresolved for now, to get more feedback.

JanKrivanek replied:
I'd strongly vote for removing code repetition.
One way to unify is to have the struct expose a common interface.
Downside of interface is the boxing/unboxing cost - however both methods doesn't feel like something that will be called in a tight loop - so that would be a very negligable cost

JanKrivanek replied:
+1 on let's get rid of the `firstManifestMediaType` arg - that's an implementation detail of the method

surayya-MS replied:
Thanks! I refactored media type validation and it looks better now. 
I also tried to use interfaces for the structs
```csharp
internal interface IPlatformSpecificManifest
{
    string mediaType { get; set; }
    long size { get; set; }
    string digest { get; set; }
    PlatformInformation platform { get; set; }
}

internal interface IImageIndex
{
    int schemaVersion { get; set; }
    string mediaType { get; set; }
    IPlatformSpecificManifest[] manifests { get; set; }
}
```
but the problem is I need to change structs 
```diff
- public record struct ManifestListV2(int schemaVersion, string mediaType, PlatformSpecificManifest[] manifests);
+ public record struct ManifestListV2(int schemaVersion, string mediaType, IPlatformSpecificManifest[] manifests) : IImageIndex;
```
same for `ImageIndexV1` struct. Because they are used in other places, probably for performance reasons, the boxing/unboxing will affect it negatively.

There is also a solution to only use `ManifestListV2`(or `ImageIndexV1`) in `ImageIndexGenerator`, that would reduce code duplication. I'm going with this route. And if in the future we would need extra fields like "annotations" in `ImageIndexV1`, I'll just change the code (not likely to happen)

SimaTian commented on src/Containers/Microsoft.NET.Build.Containers/Tasks/ComputeDotnetBaseImageAndTag.cs:
```diff
@@ -113,16 +113,37 @@ public override bool Execute()
 
+    private bool TargetRuntimeIdentiriersAreValid()
+    {
+        var muslRidsCount = TargetRuntimeIdentifiers.Count(rid => rid.StartsWith("linux-musl", StringComparison.Ordinal));
```
so we want either none or all to be linux-musl?
if ( TargetRuntimeIdentifiers[0].rid.StartsWith("linux-musl", StringComparison.Ordinal)
{
  return TargetRuntimeIdentifiers.All(rid => rid.StartsWith("linux-musl", StringComparison.Ordinal);
}
else
{
 return TargetRuntimeIdentifiers.All(rid => !rid.StartsWith("linux-musl", StringComparison.Ordinal);
}
note: I have no clue about the possible values of the muslRid or what the musl means.

surayya-MS replied:
> so we want either none or all to be linux-musl? scratch my attempt to simplify, there is too much bookkeeping for it to work nicely.

yes, exactly. 

Basically, we need this validation, because for rids starting with "linux-musl" we choose alpine image as a base image. And because we compute base image once for all rids, we check that all are "linux-musl" to get alpine image.

surayya-MS replied:
Actually, I think I can make this code shorter

surayya-MS replied:
Nope, scratch that. making the code shorter will make it confusing

JanKrivanek replied:
> Basically, we need this validation, because for rids starting with "linux-musl" we choose alpine image as a base image. And because we compute base image once for all rids, we check that all are "linux-musl" to get alpine image.

This would be a great comment in that code ;-)

SimaTian commented on src/Containers/Microsoft.NET.Build.Containers/Tasks/CreateImageIndex.cs:
```diff
@@ -0,0 +1,128 @@
+        {
+            (string imageIndex, string mediaType) = ImageIndexGenerator.GenerateImageIndex(images);
+
+            GeneratedImageIndex = imageIndex;
```
The fact that GeneratedImageIndex is declared in another file made me somewhat bamboozled since there is no other mention of it in this file as far as I could tell.
Might be just my unfamiliarity with the way Tasks are set up I suppose.
That being said, would
`(GeneratedImageIndex, string mediaType) = ...`
work in C#?

surayya-MS replied:
Yeah, I placed the `CreateImageIndex` task parameters  in `CreateImageIndex.Interface.cs` similarly to `CreateNewImage` task. But now I se there are actually not a lot of parameters, so I will just move them to one file.

> That being said, would `(GeneratedImageIndex, string mediaType) = ...` work in C#?

yes, it will. Thanks, I'll apply the change

surayya-MS replied:
done.

SimaTian commented on src/Containers/Microsoft.NET.Build.Containers/Tasks/CreateImageIndex.cs:
```diff
@@ -0,0 +1,128 @@
+        return !Log.HasLoggedErrors;
+    }
+
+    private ImageInfo[] ParseImages()
```
Few files above, we have this information. Do we have to throw it away and then parse it again or does this have a different use case?
Also, if we want to enforce the media type to be consistent, do we have to store it with each image?
Why not store it just once and then throw is we encounter mismatch (which we already kind of do)?
It is perfectly possible that I'm missing something here.

surayya-MS replied:
this is not what is happening here. 
when there is a case with multiple rids we call `_PublishMultiArchContainers` target. Inside of it, we call `_PublishSingleContainer` target per each rid, and gather outputs like manifest etc. in `GeneratedContainers` item (this is inside `_PublishSingleContainer` target). Then we call `CreateImageIndex` task with `ITaskItem[] GeneratedContainers` parameter. 
Here in `ParseImages` I parse them, and then validate the mediaType. There is no way to validate media types prior to this point.

SimaTian replied:
Thanks for the clarification

SimaTian commented on src/Containers/Microsoft.NET.Build.Containers/Tasks/CreateImageIndex.cs:
```diff
@@ -0,0 +1,128 @@
+
+    private string GetRepositoryAndTagsString()
+    {
+        _repositoryAndTagsString ??= $"{Repository}:{string.Join(", ", ImageTags)}";
```
is it possible to return the assignment directly?

surayya-MS replied:
yes, it is. I save the string because I use it twice, to save the cost of re-computing it

SimaTian commented on src/Containers/Microsoft.NET.Build.Containers/Registry/Registry.cs:
```diff
@@ -527,6 +525,17 @@ private async Task UploadBlobAsync(string repository, string digest, Stream cont
+        }          
+    }
+
     public Task PushAsync(BuiltImage builtImage, SourceImageReference source, DestinationImageReference destination, CancellationToken cancellationToken)
```
What is the difference between Push and Put in this context please?

surayya-MS replied:
it was an existing naming convention here. "Push" image - a registry method, but "Put" manifest - a manifest method (an actual put request)

surayya-MS commented on src/Tests/Microsoft.NET.Build.Containers.IntegrationTests/TargetsTests.cs:
```diff
@@ -364,6 +364,39 @@ public void MuslRidsGetAlpineContainers(string tfm, string rid, string expectedI
+    [InlineData("linux-musl-x64;linux-musl-arm64", "mcr.microsoft.com/dotnet/runtime:8.0-alpine")]
+    [InlineData("linux-x64;linux-arm64", "mcr.microsoft.com/dotnet/runtime:8.0")]
+    [Theory]
+    public void AllMuslRidsGetAlpineContainers(string rids, string expectedImage)
```
`AllOrNoneMuslRidsGetAlpineContainers` name would be misleading. All musl rids get alpine containers, and none musl rids get not alpine containers. I think the name is okay given the older test above, but I will leave this unresolved, to gather more feedback

donJoseLuis reviewed (commented):
minor comments

JanKrivanek commented on src/Containers/Microsoft.NET.Build.Containers/ImageIndexGenerator.cs:
```diff
@@ -0,0 +1,128 @@
+    /// <returns>Returns json string of image index and image index mediaType.</returns>
+    /// <exception cref="ArgumentException"></exception>
+    /// <exception cref="NotSupportedException"></exception>
+    internal static (string, string) GenerateImageIndex(ImageInfo[] imageInfos)
```
You'll be able to reuse the same type for return of `GenerateDockerManifestList` as well

surayya-MS replied:
thanks! I've decided to go with this route https://github.com/dotnet/sdk/pull/43085#discussion_r1856774326

JanKrivanek commented on src/Containers/Microsoft.NET.Build.Containers/Tasks/ComputeDotnetBaseImageAndTag.cs:
```diff
@@ -99,7 +99,7 @@ public ComputeDotnetBaseImageAndTag()
         ContainerFamily = "";
         FrameworkReferences = [];
-        TargetRuntimeIdentifier = "";
+        TargetRuntimeIdentifiers = Array.Empty<string>();
```
nit: The new empty collection initialization is more brief and expressive.

```suggestion
        TargetRuntimeIdentifiers = [];
```

JanKrivanek commented on src/Containers/Microsoft.NET.Build.Containers/Tasks/CreateImageIndex.cs:
```diff
@@ -0,0 +1,128 @@
+        var images = ParseImages();
+        if (Log.HasLoggedErrors)
+        {
+            return !Log.HasLoggedErrors;
```
```suggestion
            return false;
```

JanKrivanek reviewed (approved):
Thanks - it looks good and shippable.

I have some nonblocking comments for consideration. I'd recommend to remove code repetition and define strong return types - it'll help maintainability

surayya-MS committed:
move CreateImageIndex parameters to one class

surayya-MS committed:
use one PutAsync method instead of two

surayya-MS committed:
Merge branch 'release/8.0.4xx' into multi-arch-containers

surayya-MS committed:
Merge branch 'multi-arch-containers' of https://github.com/surayya-MS/sdk into multi-arch-containers

surayya-MS committed:
fix EndToEndMultiArch_EnvVariables

surayya-MS committed:
refactor ImageIndexGenerator manifest media type validation

surayya-MS committed:
remove lines

surayya-MS committed:
get rid of manifestMediaType arg

surayya-MS committed:
reduce code duplication when generating image index

surayya-MS commented on src/Containers/Microsoft.NET.Build.Containers/Registry/DefaultManifestOperations.cs:
```diff
@@ -41,7 +41,13 @@ public async Task<HttpResponseMessage> GetAsync(string repositoryName, string re
-        HttpContent manifestUploadContent = new StringContent(jsonString);
+
+        await PutAsync(repositoryName, reference, jsonString, mediaType, cancellationToken).ConfigureAwait(false);
+    }
```
done. thanks!

surayya-MS committed:
make GetManifestConfigDigest a prop ManifestConfigDigest

surayya-MS committed:
override ToString() for ImageInfo

surayya-MS commented on src/Containers/Microsoft.NET.Build.Containers/ImageBuilder.cs:
```diff
@@ -320,6 +320,12 @@ static string[] Split(string input)
     }
 
+    // For tests
+    internal string GetManifestConfigDigest()
```
done. thanks!

surayya-MS commented on src/Containers/Microsoft.NET.Build.Containers/ImageIndexGenerator.cs:
```diff
@@ -0,0 +1,128 @@
+    {
+        if (imageInfos.Length == 0)
+        {
+            throw new ArgumentException(string.Format(Strings.ImagesEmpty));
```
The exception message is "Cannot create manifest list (image index) because no images were provided.". `Strings` here is a class with static strings

surayya-MS commented on src/Containers/Microsoft.NET.Build.Containers/ImageIndexGenerator.cs:
```diff
@@ -0,0 +1,128 @@
+    /// <returns>Returns json string of image index and image index mediaType.</returns>
+    /// <exception cref="ArgumentException"></exception>
+    /// <exception cref="NotSupportedException"></exception>
+    internal static (string, string) GenerateImageIndex(ImageInfo[] imageInfos)
```
yes, because I'm checking for the `ManifestMediaType` of the first imageInfo to make a choice of creating docker manifest list, or oci image index. if there are no images then there is no point of creating image index

surayya-MS committed:
make manifest media type string comparison case insensitive;
small fixes;

surayya-MS committed:
fix test

surayya-MS commented on src/Containers/Microsoft.NET.Build.Containers/ImageIndexGenerator.cs:
```diff
@@ -0,0 +1,128 @@
+
+        string firstManifestMediaType = imageInfos[0].ManifestMediaType;
+
+        if (firstManifestMediaType == SchemaTypes.DockerManifestV2)
```
thanks! made it case ignore comparison

surayya-MS committed:
fix test

surayya-MS committed:
Merge branch 'release/8.0.4xx' into multi-arch-containers

surayya-MS committed:
add comment

surayya-MS committed:
Merge branch 'multi-arch-containers' of https://github.com/surayya-MS/sdk into multi-arch-containers

baronfel commented:
Very happy to merge this. Thank you for all of your work @surayya-MS, and for your deep review @SimaTian!
